File: app/src/main/java/com/lwjlol/liveeventbus/javaversion/EventLiveData.java
package com.lwjlol.liveeventbus.javaversion;

import android.os.Looper;
import android.os.SystemClock;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.collection.ArrayMap;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.OnLifecycleEvent;

import java.util.concurrent.atomic.AtomicInteger;


public class EventLiveData<T> extends MutableLiveData<T> {
    private static final Object UNSET = new Object();
    private static final Object NULL = new Object();
    private static final Object CALL = new Object();
    private static final String TAG = "EventLiveData";
    private final ArrayMap<String, Object> tempValueMap = new ArrayMap<>(2);
    private final ArrayMap<String, Observer<? super T>> foreverObserverMap = new ArrayMap<>(2);
    private final ArrayMap<String, Integer> callMap = new ArrayMap<>(2);
    private final ArrayMap<String, Boolean> isObservedMap = new ArrayMap<>();
    private final AtomicInteger callCount = new AtomicInteger(0);
    private final boolean sticky;
    @Nullable
    private Boolean lastIsCall;
    private volatile boolean invokePostValueFromCall;

    public EventLiveData(boolean sticky) {
        this.sticky = sticky;
    }


    @Nullable
    public final Boolean getLastIsCall() {
        return this.lastIsCall;
    }

    public final void call() {
        this.callCount.incrementAndGet();
        this.setCallForAll();
        this.lastIsCall = true;
        if (Looper.myLooper() == Looper.getMainLooper()) {
            super.setValue(null);
        } else {
            this.invokePostValueFromCall = true;
            super.postValue(null);
        }

    }

    public void postValue(@Nullable T value) {
        this.invokePostValueFromCall = false;
        super.postValue(value);
    }

    public void setValue(@Nullable T value) {
        this.lastIsCall = this.invokePostValueFromCall;
        this.invokePostValueFromCall = false;
        this.setValueForAll(value);
        super.setValue(value);
    }

    private void setCallForAll() {
        for (int i = 0; i < tempValueMap.size(); i++) {
            String key = tempValueMap.keyAt(i);
            tempValueMap.put(key, CALL);
        }
    }

    private void setValueForAll(@Nullable T value) {
        for (int i = 0; i < tempValueMap.size(); i++) {
            String key = tempValueMap.keyAt(i);
            if (value == null) {
                tempValueMap.put(key, NULL);
            } else {
                tempValueMap.put(key, value);
            }
        }
    }

    @MainThread
    public final void observe(@NonNull LifecycleOwner owner, @NonNull String key, @NonNull Observer<? super T> observer) {
        this.onObserve(owner, key);
        super.observe(this.get(owner), getObserverWrapper(key, observer));
    }


    public void observe(@NonNull LifecycleOwner owner, @NonNull Observer<? super T> observer) {
        this.observe(owner, this.getKey(owner), observer);
    }


    public final void observeNonNull(@NonNull LifecycleOwner owner, @Nullable String key, @NonNull NonNullObserver<? super T> block) {
        String k = key;
        if (key == null) {
            k = this.getKey(owner);
        }

        this.observe(owner, k, ((Observer<? super T>) it -> {
            if (lastIsCall != null && lastIsCall) {
                throw new IllegalArgumentException("observeNonNull unSupport observe Call, use observe");
            } else {
                if (it != null) {
                    block.onChanged(it);
                }
            }
        }));
    }

    @MainThread
    public void observeForever(@NonNull Observer<? super T> observer) {
        this.observeForever((LifecycleOwner) null, observer.toString(), observer);
    }

    @MainThread
    public final void observeForever(@Nullable LifecycleOwner owner, @Nullable String key, @NonNull Observer<? super T> observer) {
        String k = key;
        if (key == null) {
            k = this.getKey(owner);
        }
        this.onObserve(owner, k);
        Observer<? super T> observerWrapper = this.getObserverWrapper(k, observer);
        foreverObserverMap.put(key, observer);
        super.observeForever(observerWrapper);
    }

    public final void observeForeverNonNull(@Nullable LifecycleOwner owner, @Nullable String key, @NonNull NonNullObserver<? super T> block) {
        this.observeForever(owner, key, ((Observer<? super T>) it -> {
            if (lastIsCall != null && lastIsCall) {
                throw new IllegalArgumentException("observeNonNull unSupport observe Call, use observe");
            } else {
                if (it != null) {
                    block.onChanged(it);
                }
            }
        }));
    }


    public final void removeObserver(@Nullable String key, @NonNull Observer<? super T> observer) {
        super.removeObserver(observer);
        if (key == null) {
            reset(observer.toString());
        } else {
            reset(key);
        }

    }


    public void removeObserver(@NonNull Observer<? super T> observer) {
        super.removeObserver(observer);
        this.reset(observer.toString());
    }

    @NonNull
    public final LifecycleOwner get(@NonNull LifecycleOwner owner) {
        return owner instanceof Fragment && ((Fragment) owner).getView() != null ? ((Fragment) owner).getViewLifecycleOwner() : owner;
    }

    private void onObserve(LifecycleOwner owner, String key) {
        if (owner != null) {
            LifecycleOwner lifecycleOwner = this.get(owner);
            lifecycleOwner.getLifecycle().addObserver((LifecycleObserver) (new EventLiveData.OnDestroyLifecycleObserver<T>(this, key)));
        }

        (this.isObservedMap).put(key, true);
        if (this.tempValueMap.get(key) == null) {
            (this.tempValueMap).put(key, this.sticky && this.getValue() != null ? this.getValue() : UNSET);
        }

        if (lastIsCall != null && !lastIsCall && this.sticky && this.getValue() == null) {
            (this.tempValueMap).put(key, NULL);
        }

        if (lastIsCall != null && lastIsCall && this.callCount.get() > 0 && this.sticky && this.callMap.getOrDefault(key, 0) < this.callCount.get()) {
            (this.tempValueMap).put(key, CALL);
        }

    }

    private Observer<? super T> getObserverWrapper(final String key, final Observer<? super T> observer) {
        return (it -> {
            if (tempValueMap.get(key) != UNSET) {
                observer.onChanged(getValue());
                if (tempValueMap.get(key) == CALL) {
                    callMap.put(key, callCount.get());
                }
                tempValueMap.put(key, UNSET);
            }
        });
    }

    @NonNull
    public final String getKey(@Nullable LifecycleOwner owner) {
        if (owner == null) {
            return String.valueOf(SystemClock.currentThreadTimeMillis());
        }
        String k = owner.getClass().getCanonicalName();
        if (k == null) {
            k = owner.getClass().getName();
        }
        return k;
    }

    public void onClear(String key) {
        Observer<? super T> observer = this.foreverObserverMap.get(key);
        if (observer != null) {
            this.removeObserver(observer);
        } else {
            this.reset(key);
        }

    }

    @NonNull
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("\ntempValueMap = ");
        if (!tempValueMap.isEmpty()) {
            for (int i = 0; i < tempValueMap.size(); i++) {
                String key = tempValueMap.keyAt(i);
                Object value = tempValueMap.getOrDefault(key, "null");
                if (i == 0) {
                    sb.append("{");
                }
                sb.append(key).append(":").append(value);
                if (i == this.tempValueMap.size() - 1) {
                    sb.append("}");
                } else {
                    sb.append(", ");
                }
            }
        } else {
            sb.append("{}");
        }

        sb.append("\nforeverObserverMap = ");
        if (this.foreverObserverMap.keySet().isEmpty()) {
            sb.append("{}");
        } else {
            for (int i = 0; i < foreverObserverMap.size(); i++) {
                String key = foreverObserverMap.keyAt(i);
                Observer<? super T> value = foreverObserverMap.get(key);
                if (i == 0) {
                    sb.append("{");
                }
                sb.append(key).append(":");
                if (value == null) {
                    sb.append("null");
                } else {
                    sb.append(value.toString());
                }
                if (i == this.foreverObserverMap.size() - 1) {
                    sb.append("}");
                } else {
                    sb.append(", ");
                }
            }
        }

        sb.append("\nisObservedMap = ");
        if (this.isObservedMap.isEmpty()) {
            sb.append("{}");
        } else {
            for (int i = 0; i < isObservedMap.size(); i++) {
                String key = isObservedMap.keyAt(i);
                Boolean value = isObservedMap.get(key);
                if (i == 0) {
                    sb.append("{");
                }
                sb.append(key).append(":");
                if (value == null) {
                    sb.append("null");
                } else {
                    sb.append(value.toString());
                }
                if (i == this.isObservedMap.size() - 1) {
                    sb.append("}");
                } else {
                    sb.append(", ");
                }
            }
        }

        String address = super.toString();
        return address + ":\n" + sb;
    }

    private void reset(String key) {
        tempValueMap.put(key, null);
        isObservedMap.put(key, null);
        callMap.put(key, null);
        foreverObserverMap.put(key, null);
    }

    public final boolean getSticky() {
        return this.sticky;
    }

    interface NonNullObserver<T> {
        void onChanged(@NonNull T value);
    }

    static final class OnDestroyLifecycleObserver<T> implements LifecycleObserver {
        private final EventLiveData<T> livaData;
        private final String key;

        public OnDestroyLifecycleObserver(@NonNull EventLiveData<T> livaData, @NonNull String key) {
            super();
            this.livaData = livaData;
            this.key = key;
        }

        @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
        public final void onDestroy() {
            this.livaData.onClear(this.key);
        }
    }
}

File: liveeventbus/src/main/java/com/lwjlol/liveeventbus/EventLiveData.kt
package com.lwjlol.liveeventbus

import android.os.Looper
import android.os.SystemClock
import androidx.annotation.MainThread
import androidx.collection.ArrayMap
import androidx.fragment.app.Fragment
import androidx.lifecycle.*
import java.util.concurrent.atomic.AtomicInteger

/**
 * @param sticky indicate that event is a sticky event
 *
 */
class EventLiveData<T>(val sticky: Boolean = true) : MutableLiveData<T>() {
    private val tempValueMap = ArrayMap<String, Any?>(2)
    private val foreverObserverMap = ArrayMap<KeyWrapper, Observer<T>>(2)
    private val callMap = ArrayMap<String, Int>(2)
    private var isObservedMap = ArrayMap<String, Boolean>(2)
    private val callCount = AtomicInteger(0)
    private val keyMap = ArrayMap<String, ArrayList<KeyWrapper>>(2)

    /**
     * 最后一个事件是否是 [call] 发出的
     */
    var lastIsCall: Boolean? = null
        private set

    @Volatile
    private var invokePostValueFromCall = false

    /**
     * 直接发送一个 null 值来通知 [Observer] 回调，必须通过 [observe]/[observeForever] 注册才能收到结果，如果通过 [observeNonNull]/[observeForeverNonNull]
     * 注册会抛出错误。
     */
    fun call() {
        callCount.incrementAndGet()
        setCallForAll()
        lastIsCall = true
        if (Looper.myLooper() == Looper.getMainLooper()) {
            super.setValue(null)
        } else {
            invokePostValueFromCall = true
            super.postValue(null)
        }
    }

    override fun postValue(value: T?) {
        invokePostValueFromCall = false
        super.postValue(value)
    }

    override fun setValue(value: T?) {
        lastIsCall = invokePostValueFromCall
        invokePostValueFromCall = false
        setValueForAll(value)
        super.setValue(value)
    }

    private fun setCallForAll() {
        for (item in tempValueMap) {
            // 跳过非粘性还没注册的组件
            if (isObservedMap.getOrDefault(item.key, false) != true && !sticky) {
                continue
            }
            item.setValue(CALL)
        }
    }

    private fun setValueForAll(value: T?) {
        for (item in tempValueMap) {
            // 跳过非粘性还没注册的组件
            if (isObservedMap.getOrDefault(item.key, false) != true && !sticky) {
                continue
            }
            item.setValue(value ?: NULL)
        }
    }

    /**
     * @param owner
     * @param key an event only can be consumed once by same key
     * @param observer
     */
    @MainThread
    fun observe(
        owner: LifecycleOwner,
        key: String = getKey(owner),
        observer: Observer<in T>
    ) {
        val keyWrapper = KeyWrapper(key, owner, observer)
        onObserve(keyWrapper)
        super.observe(owner.get(), getObserverWrapper(key, observer))
    }

    override fun observe(
        owner: LifecycleOwner,
        observer: Observer<in T>
    ) {
        observe(owner, getKey(owner), observer)
    }

    inline fun observe(
        owner: LifecycleOwner,
        key: String? = null,
        crossinline block: ((T?) -> Unit)
    ) {
        observe(owner, key ?: getKey(owner), Observer {
            block(it)
        })
    }

    /**
     * 观察非 null 的值，对 [call] 不适用
     */
    inline fun observeNonNull(
        owner: LifecycleOwner,
        key: String? = null,
        crossinline block: ((T) -> Unit)
    ) {
        observe(owner, key ?: getKey(owner), Observer {
            require(lastIsCall == false) {
                "observeNonNull unSupport observe Call, use observe"
            }
            block(it ?: return@Observer)
        })
    }

    @MainThread
    override fun observeForever(observer: Observer<in T>) {
        observeForever(null, observer.toString(), observer)
    }

    @MainThread
    fun observeForever(
        owner: LifecycleOwner? = null,
        key: String? = null,
        observer: Observer<in T>
    ) {
        val k = key ?: getKey(owner)
        val keyWrapper = KeyWrapper(k, owner, observer)
        onObserve(keyWrapper)
        super.observeForever(getObserverWrapper(k, observer).also {
            foreverObserverMap[keyWrapper] = it
        })
    }

    inline fun observeForever(
        owner: LifecycleOwner?,
        key: String? = null,
        crossinline block: ((T?) -> Unit)
    ) {
        observeForever(owner, key, Observer {
            require(lastIsCall == false) {
                "observeNonNull unSupport observe Call, use observe"
            }
            block(it)
        })
    }

    inline fun observeForeverNonNull(
        owner: LifecycleOwner?,
        key: String? = null,
        crossinline block: ((T) -> Unit)
    ) {
        observeForever(owner, key, Observer {
            require(lastIsCall == false) {
                "observeNonNull unSupport observe Call, use observe"
            }
            block(it ?: return@Observer)
        })
    }

    private fun removeObserver(keyWrapper: KeyWrapper, observer: Observer<in T>) {
        super.removeObserver(observer)
        remove(keyWrapper)
    }

    override fun removeObserver(observer: Observer<in T>) {
        super.removeObserver(observer)
        foreverObserverMap.keys.find {
            it.observer == observer
        }?.let {
            foreverObserverMap.remove(it)
        }
    }

    fun LifecycleOwner.get() = if (this is Fragment && view != null) viewLifecycleOwner else this

    private fun onObserve(keyWrapper: KeyWrapper) {
        val key = keyWrapper.key
        val owner = keyWrapper.lifecycleOwner
        if (keyMap[key] == null) {
            keyMap[key] = ArrayList(4)
        }
        keyMap[key]?.add(keyWrapper)
        (owner?.get())?.lifecycle?.addObserver(OnDestroyLifecycleObserver(keyWrapper, this))
        isObservedMap[key] = true
        if (tempValueMap[key] == null) {
            tempValueMap[key] = if (!sticky || value == null) {
                UNSET
            } else {
                value
            }
        }
        // 最后一次是 setValue
        if (lastIsCall == false && sticky) {
            if (value == null) {
                tempValueMap[key] = NULL
            }
        }
        // 最后一次是 call 事件
        if (lastIsCall == true && callCount.get() > 0 && sticky) {
            if (callMap.getOrDefault(key, 0) < callCount.get()) {
                tempValueMap[key] = CALL
            }
        }
    }

    private fun getObserverWrapper(
        key: String,
        observer: Observer<in T>
    ): Observer<T> {
        return Observer<T> {
            if (tempValueMap[key] === UNSET) return@Observer
            @Suppress("UNCHECKED_CAST")
            observer.onChanged(value as T)
            // 消费完 call 事件就同步进度
            if (tempValueMap[key] === CALL) {
                callMap[key] = callCount.get()
            }
            tempValueMap[key] = UNSET
        }
    }

    private fun onClear(keyWrapper: KeyWrapper) {
        val observer = foreverObserverMap[keyWrapper]
        if (observer != null) {
            removeObserver(keyWrapper, observer)
        } else {
            remove(keyWrapper)
        }
    }

    private class OnDestroyLifecycleObserver<T>(
        private val keyWrapper: KeyWrapper,
        private val livaData: EventLiveData<T>
    ) : LifecycleObserver {
        @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
        fun onDestroyView() {
            livaData.onClear(keyWrapper)
        }
    }

    data class KeyWrapper(
        val key: String,
        val lifecycleOwner: LifecycleOwner? = null,
        val observer: Observer<*>? = null
    )

    override fun toString(): String {
        val sb = StringBuilder()

        sb.append("\ntempValueMap = ")
        if (tempValueMap.isNotEmpty()) {
            tempValueMap.keys.forEachIndexed { index, key ->
                if (index == 0) {
                    sb.append("{")
                }
                sb.append(key)
                    .append(":")
                    .append(tempValueMap[key])
                if (index == tempValueMap.keys.size - 1) {
                    sb.append("}")
                } else {
                    sb.append(", ")
                }
            }
        } else {
            sb.append("{}")
        }

        sb.append("\nforeverObserverMap = ")
        if (foreverObserverMap.keys.isEmpty()) {
            sb.append("{}")
        } else {
            foreverObserverMap.keys.forEachIndexed { index, key ->
                if (index == 0) {
                    sb.append("{")
                }
                sb.append(key)
                    .append(":")
                    .append(foreverObserverMap[key])
                if (index == foreverObserverMap.keys.size - 1) {
                    sb.append("}")
                } else {
                    sb.append(", ")
                }
            }
        }

        sb.append("\nisObservedMap = ")
        if (isObservedMap.isEmpty) {
            sb.append("{}")
        } else {
            isObservedMap.keys.forEachIndexed { index, key ->
                if (index == 0) {
                    sb.append("{")
                }
                sb.append(key)
                    .append(":")
                    .append(isObservedMap[key])
                if (index == isObservedMap.keys.size - 1) {
                    sb.append("}")
                } else {
                    sb.append(", ")
                }
            }
        }
        val address = super.toString()
        return "$address:\n$sb"
    }

    private fun remove(keyWrapper: KeyWrapper) {
        foreverObserverMap.remove(keyWrapper)
        keyMap[keyWrapper.key]?.remove(keyWrapper)
        if (keyMap[keyWrapper.key].isNullOrEmpty()) {
            tempValueMap.remove(keyWrapper.key)
            callMap.remove(keyWrapper.key)
            isObservedMap.remove(keyWrapper.key)
        }
    }

    companion object {
        @JvmStatic
        fun getKey(owner: LifecycleOwner?): String =
            if (owner != null) {
                owner::class.qualifiedName ?: owner::class.java.name ?: ""
            } else {
                SystemClock.currentThreadTimeMillis().toString()
            }

        private val UNSET = Any()
        private val NULL = Any()
        private val CALL = Any()
        private const val TAG = "EventLiveData"
    }
}
